File: backend/app/schemas/opportunity.py
from pydantic import BaseModel


# ============================================================
# LIST
# ============================================================


class OpportunityListItem(BaseModel):
    opportunity_id: str

    status: str
    priority: str

    part_number: str
    component_id: str

    part_name: str
    category: str
    part_family: str

    plant_code: str
    plant_name: str
    country: str

    unit_cost: float
    peer_average_cost: float

    variance_amount: float
    variance_percent: float

    potential_savings: float
    impact_score: float
    confidence_score: float


class OpportunityListResponse(BaseModel):
    items: list[OpportunityListItem]

    count: int
    limit: int
    offset: int


# ============================================================
# OPPORTUNITY DETAIL
# ============================================================


class PartDetail(BaseModel):
    part_id: str
    part_number: str
    component_id: str
    name: str
    category: str
    part_family: str
    part_type: str
    description: str | None = None


class PlantDetail(BaseModel):
    plant_id: str
    plant_code: str
    plant_name: str
    city: str
    country: str
    region: str
    currency: str


class OpportunityMetrics(BaseModel):
    unit_cost: float
    peer_average_cost: float
    variance_amount: float
    variance_percent: float

    annual_volume: int
    annual_spend: float

    potential_savings: float
    impact_score: float
    confidence_score: float


class OpportunityDetailResponse(BaseModel):
    opportunity_id: str
    opportunity_number: str

    status: str
    priority: str
    detection_source: str

    part: PartDetail
    plant: PlantDetail
    metrics: OpportunityMetrics


# ============================================================
# OVERVIEW
# ============================================================


class CostDriver(BaseModel):
    code: str
    name: str

    impact_amount: float
    impact_percent: float

    rank: int
    confidence_score: float

    explanation: str


class OverviewResponse(BaseModel):
    opportunity_id: str

    metrics: OpportunityMetrics

    cost_drivers: list[CostDriver]


# ============================================================
# PLANTS
# ============================================================


class PlantComparison(BaseModel):
    plant_id: str
    plant_code: str
    plant_name: str
    country: str

    unit_cost: float
    peer_average_cost: float

    variance_amount: float
    variance_percent: float

    annual_volume: int
    volume_variance_percent: float

    rank: int


class PlantsResponse(BaseModel):
    opportunity_id: str

    benchmark_type: str

    peer_average_cost: float

    plants: list[PlantComparison]


# ============================================================
# SUPPLIERS
# ============================================================


class SupplierComparison(BaseModel):
    supplier_id: str
    supplier_code: str
    supplier_name: str
    country: str

    unit_cost: float
    peer_average_cost: float

    variance_amount: float
    variance_percent: float

    annual_spend: float
    annual_volume: int

    quality_score: float
    delivery_score: float
    responsiveness_score: float
    overall_score: float

    rank: int


class SuppliersResponse(BaseModel):
    opportunity_id: str

    suppliers: list[SupplierComparison]


# ============================================================
# LOGISTICS
# ============================================================


class LogisticsComponent(BaseModel):
    code: str
    name: str

    cost: float
    peer_average: float

    variance: float
    variance_percent: float

    rank: int


class LogisticsTrendPoint(BaseModel):
    period: str

    actual_cost: float
    peer_average_cost: float


class LogisticsResponse(BaseModel):
    opportunity_id: str

    total_cost: float
    peer_average: float

    variance: float
    variance_percent: float

    components: list[LogisticsComponent]

    trend: list[LogisticsTrendPoint]


# ============================================================
# TARIFF
# ============================================================


class TariffPlantComparison(BaseModel):
    plant_id: str
    plant_code: str
    plant_name: str
    country: str

    duty_rate: float


class TariffResponse(BaseModel):
    opportunity_id: str

    hs_code: str

    duty_rate: float
    peer_average_duty_rate: float

    calculation_basis: str
    valuation_type: str
    effective_date: str

    import_duty_per_unit: float
    peer_duty_per_unit: float

    duty_variance_per_unit: float
    duty_variance_percent: float

    annual_duty_impact: float

    plant_comparisons: list[TariffPlantComparison]

File: frontend/src/types/opportunity.ts
// ==========================================================
// OPPORTUNITY LIST
// ==========================================================

export interface OpportunityListItem {
  opportunity_id: string;

  status: string;
  priority: string;

  part_number: string;
  component_id: string;

  part_name: string;
  category: string;
  part_family: string;

  plant_code: string;
  plant_name: string;
  country: string;

  unit_cost: number;
  peer_average_cost: number;

  variance_amount: number;
  variance_percent: number;

  potential_savings: number;
  impact_score: number;
  confidence_score: number;
}


export interface OpportunityListResponse {
  items: OpportunityListItem[];

  count: number;
  limit: number;
  offset: number;
}


// ==========================================================
// RECSYS
// ==========================================================

export interface RankedOpportunity {
  opportunity_id: string;

  part_number: string;
  part_name: string;

  plant_code: string;
  plant_name: string;
  country: string;

  potential_savings: number;
  confidence_score: number;

  final_score: number;
  rank: number;

  reasons: string[];
}


export interface RankingResponse {
  run_id: string | null;

  candidate_count: number;
  eligible_count: number;
  ranked_count: number;

  items: RankedOpportunity[];
}


// ==========================================================
// COMMON OPPORTUNITY METRICS
// ==========================================================

export interface OpportunityMetrics {
  unit_cost: number;
  peer_average_cost: number;

  variance_amount: number;
  variance_percent: number;

  annual_volume: number;
  annual_spend: number;

  potential_savings: number;

  impact_score: number;
  confidence_score: number;
}


// ==========================================================
// OPPORTUNITY DETAIL
// ==========================================================

export interface PartDetail {
  part_id: string;

  part_number: string;
  component_id: string;

  name: string;
  category: string;
  part_family: string;
  part_type: string;

  description: string | null;
}


export interface PlantDetail {
  plant_id: string;

  plant_code: string;
  plant_name: string;

  city: string;
  country: string;
  region: string;

  currency: string;
}


export interface OpportunityDetailResponse {
  opportunity_id: string;
  opportunity_number: string;

  status: string;
  priority: string;

  detection_source: string;

  part: PartDetail;
  plant: PlantDetail;

  metrics: OpportunityMetrics;
}


// ==========================================================
// OVERVIEW
// ==========================================================

export interface CostDriver {
  code: string;
  name: string;

  impact_amount: number;
  impact_percent: number;

  rank: number;
  confidence_score: number;

  explanation: string;
}


export interface OverviewResponse {
  opportunity_id: string;

  metrics: OpportunityMetrics;

  cost_drivers: CostDriver[];
}


// ==========================================================
// PLANT COMPARISON
// ==========================================================

export interface PlantComparison {
  plant_id: string;

  plant_code: string;
  plant_name: string;
  country: string;

  unit_cost: number;
  peer_average_cost: number;

  variance_amount: number;
  variance_percent: number;

  annual_volume: number;
  volume_variance_percent: number;

  rank: number;
}


export interface PlantsResponse {
  opportunity_id: string;

  benchmark_type: string;

  peer_average_cost: number;

  plants: PlantComparison[];
}


// ==========================================================
// SUPPLIER COMPARISON
// ==========================================================

export interface SupplierComparison {
  supplier_id: string;

  supplier_code: string;
  supplier_name: string;

  country: string;

  unit_cost: number;
  peer_average_cost: number;

  variance_amount: number;
  variance_percent: number;

  annual_spend: number;
  annual_volume: number;

  quality_score: number;
  delivery_score: number;
  responsiveness_score: number;
  overall_score: number;

  rank: number;
}


export interface SuppliersResponse {
  opportunity_id: string;

  suppliers: SupplierComparison[];
}


// ==========================================================
// LOGISTICS
// ==========================================================

export interface LogisticsComponent {
  code: string;
  name: string;

  cost: number;
  peer_average: number;

  variance: number;
  variance_percent: number;

  rank: number;
}


export interface LogisticsTrendPoint {
  period: string;

  actual_cost: number;
  peer_average_cost: number;
}


export interface LogisticsResponse {
  opportunity_id: string;

  total_cost: number;
  peer_average: number;

  variance: number;
  variance_percent: number;

  components: LogisticsComponent[];

  trend: LogisticsTrendPoint[];
}


// ==========================================================
// TARIFF
// ==========================================================

export interface TariffPlantComparison {
  plant_id: string;

  plant_code: string;
  plant_name: string;
  country: string;

  duty_rate: number;
}


export interface TariffResponse {
  opportunity_id: string;

  hs_code: string;

  duty_rate: number;
  peer_average_duty_rate: number;

  calculation_basis: string;
  valuation_type: string;

  effective_date: string;

  import_duty_per_unit: number;
  peer_duty_per_unit: number;

  duty_variance_per_unit: number;
  duty_variance_percent: number;

  annual_duty_impact: number;

  plant_comparisons: TariffPlantComparison[];
}